xtask/src/run.rs

```diff
@@ -83,40 +102,383 @@ pub fn build(opts: BuildOptions) -> Result<Vec<(String, PathBuf)>> {
+                        writeln!(&mut failures, "{profile}:{name} failed: {status:?}")
+                            .context("String write failed")?
+                    }
+                }
+            }
+            if failures.is_empty() {
+                Ok(())
+            } else {
+                Err(anyhow!("failures:\n{}", failures))
+            }
+        }
+        Environment::VM { kernel_image } => {
+            let cache_dir = Path::new("test/.tmp");
+            create_dir_all(cache_dir).context("failed to create cache dir")?;
+            let gen_init_cpio = cache_dir.join("gen_init_cpio");
+            if !gen_init_cpio
+                .try_exists()
+                .context("failed to check existence of gen_init_cpio")?
+            {
+                let mut curl = Command::new("curl");
+                curl.args([
+                    "-sfSL",
+                    "https://raw.githubusercontent.com/torvalds/linux/master/usr/gen_init_cpio.c",
+                ]);
+                let mut curl_child = curl
+                    .stdout(Stdio::piped())
+                    .spawn()
+                    .with_context(|| format!("failed to spawn {curl:?}"))?;
+                let Child { stdout, .. } = &mut curl_child;
+                let curl_stdout = stdout.take().unwrap();
+
+                let mut clang = Command::new("clang");
+                exec(
+                    clang
+                        .args(["-g", "-O2", "-x", "c", "-", "-o"])
+                        .arg(&gen_init_cpio)
+                        .stdin(curl_stdout),
+                )
+                .with_context(|| format!("failed to run {clang:?}"))?;
+
+                let output = curl_child
+                    .wait_with_output()
+                    .with_context(|| format!("failed to wait for {curl:?}"))?;
+                let Output { status, .. } = &output;
+                if status.code() != Some(0) {
+                    bail!("{curl:?} failed: {output:?}")
+                }
+            }
+
+            let mut errors = Vec::new();
+            for kernel_image in kernel_image {
+                // Guess the guest architecture.
+                let mut cmd = Command::new("file");
+                let output = cmd
+                    .arg("--brief")
+                    .arg(&kernel_image)
+                    .output()
+                    .with_context(|| format!("failed to run {cmd:?}"))?;
+                let Output { status, .. } = &output;
+                if status.code() != Some(0) {
+                    bail!("{cmd:?} failed: {output:?}")
+                }
+                let Output { stdout, .. } = output;
+
+                // Now parse the output of the file command, which looks something like
+                //
+                // - Linux kernel ARM64 boot executable Image, little-endian, 4K pages
+                //
+                // - Linux kernel x86 boot executable bzImage, version 6.1.0-10-cloud-amd64 [..]
```

Comment:
This works, but I find it hard to believe we have to call `file` to get this info.
I'm sure there's another way, but it's not hit me just yet

Reply:
Thinking about this again, do we want/need to support a different guest arch to the host arch?

Reply:
If it's useful we keep it.